### PR TITLE
feat(medusa, stock-location, inventory): Allow modules to integrate with core

### PR DESCRIPTION
**Pull request by carlos-r-l-rodrigues**
What:
Enhanced Module's configs enabling them to share the Core's EntityManager.
Modules will be able to run integrated in the Core and have the option to use their own resources like Database, or use the same as the Core.

FIXES: CORE-980

### CODE REVIEW

File: packages/inventory/src/loaders/connection.ts
Comment (srindom) [inline packages/inventory/src/loaders/connection.ts:17]: thought: right now the module would not work without being in a shared scenario so I am wondering if we should remove the connection loader altogether first time around?

Comment (srindom) [reply packages/inventory/src/loaders/connection.ts:17]: My point is that we should leave a potential code path open that might not work. Wdyt?

File: packages/inventory/src/services/inventory-item.ts
Comment (srindom) [inline packages/inventory/src/services/inventory-item.ts:224]: thought: if the module is not registered as a shared module this would break. The reason is the following:

- core manager creates transaction
- `inventoryService.withTransaction(coreManager).something()`
- now InventoryService is trying to use a manager that doesn't have the models from the module registered so can't do e.g. `manager.getRepository(InventoryItem)`

Alternatively, if the core doesn't call the inventoryService.withTransaction the following happens:

- InventoryService calls atomicPhase which creates a new transaction using the InventoryService connection.
- the transactionManager created by atomicPhase doesn't know about the StagedJob model so calling `eventBusService.withTransaction(inventoryManager)` will fail.

Again I think it might make sense, the first time around to only support the shared approach and not leave any code paths open that cannot work.

Comment (srindom) [reply packages/inventory/src/services/inventory-item.ts:224]: A thing to be aware of in any case is that when we start looking at distributing these this would be a thing that would need refactoring of some sort.

**Review by srindom — COMMENTED** [review]
Think this is great - feels like a good first approach. Would like to understand how we will approach the TO as well.

File: packages/inventory/src/loaders/connection.ts
Comment (carlos-r-l-rodrigues) [reply packages/inventory/src/loaders/connection.ts]: That's a good point.
I'll throw an Error at the constructor if the resource is not shared until we have eventServices as a Module

File: packages/inventory/src/services/inventory-item.ts
Comment (carlos-r-l-rodrigues) [reply packages/inventory/src/services/inventory-item.ts:224]: Good catch. Just refactored it a bit and tested the flow manually to check.
I'm restricting the modules to be only be used with shared resources.
When EventBus become a Module we refactor modules using it.

File: packages/inventory/src/services/inventory.ts
Comment (adrien2p) [inline packages/inventory/src/services/inventory.ts:55]: suggestion: `if (moduleDeclaration?.resources !== MODULE_RESOURCE_TYPE.SHARED)` then throw as it is not supported at all for now.

File: packages/inventory/src/services/inventory.ts
Comment (adrien2p) [inline packages/inventory/src/services/inventory.ts:66]: thought: It is a shame that we are losing the injection for that service. It is not a good practice to instantiate a class from the constructor as it hides those and creates unnecessary complexity to test. I would have thought that the module would have at least its own container to register all those services. Would it make sense? something like 
`createContainer({ something }, coreContainer)` but since all resources are shared should we export all the services for the core to register them? what is you opinion?

File: packages/medusa/src/loaders/module-definitions/index.ts
Comment (adrien2p) [inline packages/medusa/src/loaders/module-definitions/index.ts:44]: todo: null is an object too. should we should check that

File: packages/medusa/src/loaders/module.ts
Comment (adrien2p) [inline packages/medusa/src/loaders/module.ts:57]: todo: on a class you can get Class.name or on an instance Class.constructor.name

File: packages/medusa/src/types/global.ts
Comment (adrien2p) [inline packages/medusa/src/types/global.ts:52]: suggestion: should we set it to shared as of now?

**Review by adrien2p — CHANGES_REQUESTED** [review]

File: packages/inventory/src/services/inventory.ts
Comment (srindom) [reply packages/inventory/src/services/inventory.ts:66]: The idea here is that the `InventoryItemService`, etc. are meant to be hidden from the global container and only accessible through one API (namely, the `IInventoryServce`). A separate container for the module could make sense - although I am not sure how we would then handle the InventoryService which would need injections from both the "core container" and the module container.

File: packages/medusa/src/loaders/module.ts
Comment (srindom) [inline packages/medusa/src/loaders/module.ts:65]: thought: could make sense to not allow resolution of the models directly. This would make it harder to abuse the models, but still make it available in `db_entities` for the core connection.

File: packages/inventory/src/services/inventory.ts
Comment (adrien2p) [reply packages/inventory/src/services/inventory.ts:66]: we can do something like that:
The loader already accept the container from the core in the options:
- Create a local container that take the core container as the second argument
- register the different services from the module
- resolve the `InventoryItemService`
- register it

```ts
const container = createContainer({}, coreContainer)
container.register("serviceA", asFunction((cradle) => new ServiceA(cradle)))
container.register("serviceB", asFunction((cradle) => new ServiceB(cradle))) // this service is injected into A

const serviceA = container.resolve("serviceA")

// either that
coreContainer.register("serviceA", asValue(serviceA))
// or another approach

```

what do you think? I did not test it, it is just pseudo code

Comment (adrien2p) [reply packages/inventory/src/services/inventory.ts:66]: It is very similar to what I do in the tests

File: packages/medusa/src/loaders/module.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/loaders/module.ts]: That is a good point and will force the Module to be only accessible through the service layer

File: packages/medusa/src/types/global.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/types/global.ts:52]: I think is OK to keep it here. The 2 modules we have aren't supporting isolated resources for now, but that doesn't mean others couldn't.

File: packages/medusa/src/loaders/module.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/loaders/module.ts]: Thanks. constructor.name was returning Function and I quickly added this to test.
.name does the trick, but I will remove all that to do not register models to the core container

File: packages/medusa/src/loaders/module-definitions/index.ts
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/loaders/module-definitions/index.ts:44]: that's ok, the spread operator will ignore it
```
moduleDeclaration: {
  ...definition.defaultModuleDeclaration,
  ...moduleDeclaration,
},
```

File: packages/inventory/src/services/inventory.ts
Comment (carlos-r-l-rodrigues) [reply packages/inventory/src/services/inventory.ts:62]: I don't think the core should import more than what is expected. These 2 modules will have their own container.
They don't have it yet because the tests are not in place. Next steps will be integrate the modules in our integration-tests, check the best way to add integration tests in these modules as well + the unit tests

Comment (srindom) [reply packages/inventory/src/services/inventory.ts:62]: I am fine with holding off with module containers for now.

Comment (adrien2p) [reply packages/inventory/src/services/inventory.ts:62]: I am fine too for now :D

**Review by srindom — APPROVED** [review]
LGTM!

File: packages/stock-location/src/services/stock-location.ts
Comment (olivermrbl) [inline packages/stock-location/src/services/stock-location.ts:239]: thought: I believe `findOne` returns the first row in the table, if `undefined` is passed. Should we add 1) a check for defined `addressId`, as we do in the core, or 2) use a `where { id: addressId }` config, as this will work as expected?

**Review by olivermrbl — APPROVED** [review]
LGTM, just a single thought :)

File: packages/stock-location/src/services/stock-location.ts
Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts]: Good catch. undefined has to be validated even using where.
That also reminded me there are actually multiple places where `findOne` with optional relations should be replaced by `find` to prevent typeOrm from performing two queries

**Comment by olivermrbl** [comment]
@adrien2p When you have time, can I get you to run another review, so we get closer to merging?

File: packages/inventory/src/services/inventory-item.ts
Comment (adrien2p) [inline packages/inventory/src/services/inventory-item.ts:51]: suggest: I think that it should be part of a custom method in the repository. Not sure it should stay in the service.
Usually, here you use the buildQuery, pass everything to the custom method repo and then you can manage the location and stuff but half of it is already managed by the utility. wdyt?

File: packages/stock-location/src/services/stock-location.ts
Comment (adrien2p) [inline packages/stock-location/src/services/stock-location.ts:256]: question: Once merged, the metadata has already been merged as well isn't it?

**Review by adrien2p — COMMENTED** [review]
LGTM: one question and one point:
I really don't think that we should start putting repository specificities into the service. Having to manage the query builder and relations from the service seams a bit off. I let you look at the comment and let me know what you think.

File: packages/stock-location/src/services/stock-location.ts
Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts:256]: typeorm `merge` doesn't have the same behavior `setMetadata` has.

Comment (adrien2p) [reply packages/stock-location/src/services/stock-location.ts:256]: I understand that, wanted you opinion on the global approach. Does it make sense to merge the metadata and the re copying them all. I would say that we can extract the metadata before the merge and then call the setMetada to avoid any side effects. Wdyt?

**Comment by olivermrbl** [comment]
@adrien2p Please use "Approve" if the current state LGTY :)

**Review by adrien2p — APPROVED** [review]

**Comment by adrien2p** [comment]
@carlos-r-l-rodrigues if you are alright with the repository specific work in the service for that pr then you can merge and tackle that later. Just want to avoid accumulating debt across pr as it is easily forgotten. up to you

File: packages/inventory/src/services/inventory-item.ts
Comment (carlos-r-l-rodrigues) [reply packages/inventory/src/services/inventory-item.ts]: For now I will just move it to a helper function. This module is sharing resources with core, and custom repositories should also be exported to be loaded, I have to test all that.

File: packages/stock-location/src/services/stock-location.ts
Comment (carlos-r-l-rodrigues) [reply packages/stock-location/src/services/stock-location.ts:262]: fixed

**Pull request merged**
